### PR TITLE
Send events to queue

### LINKED ISSUE #1007: Refactor webhooks to use a queue system
# What is needed?
Don't send webhooks from the transaction service. Send them to a queue, so services can subscribe and get the needed webhooks.

# Background
- Current webhook implementation requires the tx service to POST to the configured servers.
- Some servers can take some time to answer, blocking the tx service (a quick solution could be reducing the `timeout`).
- Some third party services would like to integrate with our webhooks, but that could impact the performance of the tx service.

# Proposed solution
- Send webhooks to a queue (rabbitmq, kafka...)
- A new service (webhook service) reads the webhooks and sends them to 3rd party clients
- From that queue, CGW can also read the webhooks

# Solution
- The queue chosen is RabbitMq
- The queues are going to be created by the safe-events-service
- The publisher is going to publish to an exchange in `FanOut` mode 
- The publisher class should be a singleton class that keeps only one connection open per process
- One RabbitMq for all chains

# Schema
Queues-rabbit-mq(2)

### PR DESCRIPTION
**Pull request by moisses89**
Closes #1007 
- add class to manage asynchronous connection (`gevent`) with `rabbitmq`.
- add class to manage synchronous connection for testing purposes. 
- add mock class to don't connect when there is not `rabbitmq` host configured.
- add singleton provider class to keep only one connection opened.

### CODE REVIEW

File: config/settings/test.py [hunk 1]
```diff
@@ -47,3 +47,5 @@
         "level": "DEBUG",
     }
 }
+
+EVENTS_QUEUE_ASYNC_CONNECTION = False
```
Comment (moisses89) [inline config/settings/test.py:51]: Needed to force to use blockingconnection when testing as we can test with geventconnection

File: safe_transaction_service/history/signals.py [hunk 2]
```diff
@@ -154,19 +156,22 @@ def process_webhook(
...
-                )
-                send_webhook_task.apply_async(
-                    args=(address, payload), priority=2  # Almost lowest priority
-                )  # Almost the lowest priority
-                send_notification_task.apply_async(
-                    args=(address, payload),
-                    countdown=5,
-                    priority=2,  # Almost lowest priority
-                )
+                if mark_notification_as_processed(address, payload):
```
Comment (moisses89) [inline safe_transaction_service/history/signals.py:159]: @Uxio0 I move the check if payload is repeated here, at this way we could avoid repeated webhooks and repeated notifications, but maybe is in notifications because some reason that I missing.

File: safe_transaction_service/events/services/queue_service.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    EXCHANGE_NAME = settings.EVENTS_QUEUE_RABBITMQ_EXCHANGE_NAME
+
+    def __init__(self):
+        self._channel = None
+        self._connection: GeventConnection = None
+        self.unsent_events = []
+        credentials = PlainCredentials(
+            settings.EVENTS_QUEUE_USERNAME, settings.EVENTS_QUEUE_PASSWORD
+        )
+        self._connection_parameters = ConnectionParameters(
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:45]: I think using URL parameters is way more flexible. I would only leave password out due to security reasons: https://pika.readthedocs.io/en/stable/modules/parameters.html#urlparameters

Comment (moisses89) [reply safe_transaction_service/events/services/queue_service.py]: Done!

File: safe_transaction_service/history/signals.py (hunk 2 above)
Comment (Uxio0) [reply safe_transaction_service/history/signals.py:159]: I think it's a good idea but I would not change it on this PR, as it can break the current behaviour. But if you already refactored the tests don't worry and we will merge it like this

Comment (moisses89) [reply safe_transaction_service/history/signals.py:159]: I did not change the tests, but the tests still works. 
I'm concerned about repeated events in the queue.
If we don't want to change the code here, maybe I can call `is_duplicated` method before to call `send_notification` (because send_notification is setting the key) If I do this after `send_notification`  then the event will be send never.

Comment (moisses89) [reply safe_transaction_service/history/signals.py]: Anyway, we can handle it in a different PR. 
@Uxio0 I removed the commit that included the mentioned changes.

File: .github/workflows/python.yml [hunk 4]
```diff
@@ -86,6 +95,7 @@ jobs:
         ETHEREUM_TRACING_NODE_URL: http://localhost:8545
         ETH_HASH_BACKEND: pysha3
         REDIS_URL: redis://localhost:6379/0
+        EVENTS_QUEUE_URL_HOST: amqp://guest:guest@localhost:5672/
```
Comment (Uxio0) [inline .github/workflows/python.yml:98]: EVENTS_QUEUE_URL_HOST -> EVENTS_QUEUE_URL

File: config/settings/base.py [hunk 5]
```diff
@@ -489,6 +494,13 @@
     "ALERT_OUT_OF_SYNC_EVENTS_THRESHOLD", default=0.1
 )  # Percentage of Safes allowed to be out of sync without alerting. By default 10%
 
+# Events
+# ------------------------------------------------------------------------------
+EVENTS_QUEUE_URL_HOST = env("EVENTS_QUEUE_URL_HOST", default=None)
+EVENTS_QUEUE_ASYNC_CONNECTION = env("EVENTS_QUEUE_ASYNC_CONNECTION", default=False)
+EVENTS_QUEUE_RABBITMQ_EXCHANGE_NAME = env(
```
Comment (Uxio0) [inline config/settings/base.py:501]: EVENTS_QUEUE_RABBITMQ_EXCHANGE_NAME -> EVENTS_QUEUE_EXCHANGE_NAME

File: run_tests.sh [hunk 6]
```diff
@@ -4,9 +4,9 @@ set -euo pipefail
 
 export DJANGO_SETTINGS_MODULE=config.settings.test
 export DJANGO_DOT_ENV_FILE=.env.test
-docker compose -f docker-compose.yml -f docker-compose.dev.yml build --force-rm db redis ganache
-docker compose -f docker-compose.yml -f docker-compose.dev.yml up --no-start db redis ganache
-docker compose -f docker-compose.yml -f docker-compose.dev.yml start db redis ganache
+docker compose -f docker-compose.yml -f docker-compose.dev.yml build --force-rm db redis ganache rabbitmq
+docker compose -f docker-compose.yml -f docker-compose.dev.yml up --no-start db redis ganache rabbitmq
+docker compose -f docker-compose.yml -f docker-compose.dev.yml start db redis ganache rabbitmq
```
Comment (Uxio0) [inline run_tests.sh:9]: Maybe it's time to add healthcheck to docker compose now that's supported and remove the sleep: https://docs.docker.com/compose/compose-file/05-services/#healthcheck (we can do it in another PR)

File: safe_transaction_service/events/services/queue_service.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    def __new__(cls):
+        if not hasattr(cls, "instance"):
+            if settings.EVENTS_QUEUE_URL_HOST:
+                if settings.EVENTS_QUEUE_ASYNC_CONNECTION:
+                    cls.instance = AsyncQueueService()
+                else:
+                    cls.instance = SyncQueueService()
+            else:
+                # Mock send_event to not configured host us is not mandatory configure a queue for events
+                cls.instance = MockedQueueService()
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:26]: Log a warning if a `Mocked` service is used

File: safe_transaction_service/events/services/queue_service.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        return cls.instance
+
+    @classmethod
+    def del_singleton(cls):
+        if hasattr(cls, "instance"):
+            del cls.instance
+
+
+class QueueService:
+    EXCHANGE_NAME = settings.EVENTS_QUEUE_RABBITMQ_EXCHANGE_NAME
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:36]: Set this inside `__init__` (or pass it as a parameter in __init__ and send it on the provider)

File: safe_transaction_service/events/services/queue_service.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+    def __init__(self):
+        self._channel = None
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:39]: Type

File: safe_transaction_service/events/services/queue_service.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        self._connection: GeventConnection = None
+        self.unsent_events = []
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:41]: Type

File: safe_transaction_service/events/services/queue_service.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+        self._connection_parameters = URLParameters(settings.EVENTS_QUEUE_URL_HOST)
+
+    def send_event(
+        self, payload: Dict[str, Any], fail_retry: Optional[bool] = True
+    ) -> bool:
+        """
+        Send an event to rabbitMq exchange
+
+        :param event: Dict with the payload of the event
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:50]: Params missing on the doc and not matching the function

File: safe_transaction_service/events/services/queue_service.py [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+            return False
+
+        try:
+            event = json.dumps(payload)
+            self._channel.basic_publish(
+                exchange=self.EXCHANGE_NAME, routing_key="", body=event
+            )
+            return True
+        except pika.exceptions.ConnectionClosedByBroker:
+            logger.warning("Event can not be sended due there is no channel opened")
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:65]:
```suggestion
            logger.warning("Event can not be sent due to there is no channel opened")
```

File: safe_transaction_service/events/services/queue_service.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+            self._connection_parameters,
+            on_open_callback=self.on_connection_open,
+            on_open_error_callback=self.on_connection_open_error,
+            on_close_callback=self.on_connection_closed,
+        )
+
+    def on_connection_open(self, connection: GeventConnection):
+        """This method is called by pika once the connection to RabbitMQ has
+        been established. It passes the handle to the connection object.
+        :param GeventConnection connection: The connection
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:114]:
```suggestion
        """
        This method is called by pika once the connection to RabbitMQ has
        been established. It passes the handle to the connection object.
        
        :param GeventConnection connection: The connection
```

File: safe_transaction_service/events/services/queue_service.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+class SyncQueueService(QueueService):
+    """
+    Synchronous connection with test purpose as we cannot test using gevent connection
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.connect()
+
+    def connect(self):
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:233]: I think you should return the connection when connecting, even if you are assigning locally

File: safe_transaction_service/events/services/queue_service.py [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+        )
+
+
+class MockedQueueService:
+    """
+    Mocked class to use in case that there is not rabbitMq queue to send events
+    """
+
+    def send_event(self, event: Dict[str, Any]):
+        logger.warning("MockedQueueService: Not sending event with payload %s", event)
```
Comment (Uxio0) [inline safe_transaction_service/events/services/queue_service.py:272]: I would make this debug, as people with this not configured should not get a warning

**Review by Uxio0 — APPROVED** [review]
Really nice job!
